### DISCUSSION TITLE
Histogram: Domain-Axis shows non-exising-values

### DISCUSSION THREAD

**Discussion by JP95Git** [jfree/jfreechart#283] [Q&A]
Hi,

I am using jFreeChart 1.5.3 and create histograms with a code like this

```
protected void doGet(HttpServletRequest request, HttpServletResponse response) {

	// Load the chart-data from database
	// Range of values: ..., -0.2, -0.1, 0, 0.1, 0.2, ...
	HistogramDataset dataset = getHistogramData(connection, ...);

	OutputStream out = response.getOutputStream();

	JFreeChart chart = ChartFactory.createHistogram("", "domainAxis", "rangeAxis", dataset, PlotOrientation.VERTICAL, true, false, false);
	XYPlot plot = (XYPlot) chart.getPlot();

	NumberAxis domainAxis = (NumberAxis) plot.getDomainAxis();
	domainAxis.setAxisLineVisible(false);

	ValueAxis rangeAxis = plot.getRangeAxis();
	rangeAxis.setAxisLineVisible(false);

	ChartUtils.writeChartAsPNG(out, chart, 800, 600);
}
```
This works fine, if the chart if based on many values (left image), but didn't look so good if I have fewer values (right image).

![chart-report](https://user-images.githubusercontent.com/78929417/156817325-540fe9dd-572d-484d-9e1f-91276e005a00.png)

Furthermore I see some problems with the right chart:
- The domain-axis shows values like 0.12, which do not exists in the source-data.
- It is not clear which value the tiny lines between the bars are, e.g. between 0.24 and 0.26

**Comment by trashgod** [comment]
Your chosen `ChartFactory` uses an [`XYBarRenderer`](https://www.jfree.org/jfreechart/javadoc/org/jfree/chart/renderer/xy/XYBarRenderer.html) as its `XYItemRenderer`. You can adjust the bars' _margin_ as suggested [here](https://stackoverflow.com/a/22980115/230513) and [here](https://stackoverflow.com/a/7407471/230513).

**Reply by JP95Git** [reply]
Thanks, but the margin is not my problem. I added it for testing purposes.

I made a quick comparison with jFreeChart (left) and Microsoft Excel (right), same data:
![grafik](https://user-images.githubusercontent.com/78929417/157058790-c3f58b0e-e399-4289-9fdd-3e9167e4980d.png)

In Excel you can see only values on the domain-axis, which actually exists in the data, in jFreeChart you see a lot of values on the domain-axis, which are not in the data. This is fine, if there is not enough room to show all existing values on the domain-axis, but if there is enough room for all values, only these should be shown.

So it would be great to have 
- a chart like the top left, if there is not enough room for each value on the domain-axis.
- a chart like the bottom right, if there is enough room for each value on the domain-axis.

**Comment by trashgod** [comment]
You might also look at `setBarAlignmentFactor()` or custom tick units on the chart's domain axis. Alternatively, consider a `CategoryPlot` having a domain `CategoryAxis`.

**Comment by JP95Git** [comment]
Thanks. I can't test this at the moment because I have no time left for this project. I already thought about custom tick units, since they solved some issues with other charts. CategoryPlot sounds promising, just by its name.

But... I noticed this some minutes ago:
The chart created by jFreeChart is incorrect, since we have only the values 0.2, 0.3 and 0.4 in the data. While the left and middle bar are somehow around 0.2 and 0.3, the 0.4 has no bar, because the right bar is from 0.34 to 0.39...  but not 0.4. So this is clearly a bug for me, Didn't noticed this until I created the chart with Excel.

Here is the used data as csv:
[histogram-bug.csv](https://github.com/jfree/jfreechart/files/8199785/histogram-bug.csv)

**Comment by trashgod** [comment]
Using the example shown in issue #288, the image below compares two approaches:

- Above, a histogram chart using a `HistogramDataset` having custom bounds:

        dataset.addSeries("data", values, 3, .15, .45);

- Below, a bar chart using a `CategoryDataset` summarizing the same `HistogramDataset`:

        NumberFormat nf = NumberFormat.getNumberInstance();
        String rowKey = "data";
        DefaultCategoryDataset catgoryDataset = new DefaultCategoryDataset();
        for (int i = 0; i < dataset.getItemCount(0); i++) {
            int count = dataset.getY(0, i).intValue();
            String colKey = nf.format(dataset.getX(0, i));
            catgoryDataset.addValue(count, rowKey, colKey);
        }

![HistogramCategory](https://user-images.githubusercontent.com/32623871/158527984-124e2064-6a91-402e-8dd8-a372ec40ec02.png)

You might also look at a `SimpleHistogramDataset` with custom `SimpleHistogramBin` bounds, seen [here](https://stackoverflow.com/a/29283528/230513).
